### ISSUE TITLE
any plan to support delta repo updates? 

### ISSUE THREAD

**Issue by rezad1393** [NeoApplications/Neo-Store#455]
https://f-droid.org/en/2023/03/01/new-repo-format-faster-smaller-updates.html

**Comment by machiav3lli** [comment]
Yup, most probably after 1.0.0

**Comment by 48-mq** [comment]
Any update on this?

**Comment by grote** [comment]
also relevant here: https://f-droid.org/2023/05/02/three-client-libraries.html

**Comment by machiav3lli** [comment]
As said it's on the TODO-list, but first the main focus of 1.1.X will be IndexV2 and possibly delta will come along.

**Comment by 48-mq** [comment]
Thanks @machiav3lli I will check back in a couple of months :)

**Comment by machiav3lli** [comment]
Support added in 1.1.0-beta01

**Comment by grote** [comment]
great news! I hope you have good test coverage for the diffing now. F-Droid client just had issues with it again that needed manual intervention on the repo side to fix: https://f-droid.org/en/2025/07/17/twif.html

**Comment by machiav3lli** [comment]
@grote For this (second) implementation, I went on with a more generalizable patcher (than the one I did earlier on Droidify), which I'd PR later to Droidify. Here's the patcher if you'd like to take a look: [com.machiav3lli.fdroid.data.index.v2.IndexV2Merger](https://github.com/NeoApplications/Neo-Store/blob/master/src/main/kotlin/com/machiav3lli/fdroid/data/index/v2/IndexV2Merger.kt)

**Comment by grote** [comment]
Are you verifying the signature of `entry.jar` and the hashes of the files in there? 

Then it looks like you keep a full index JSON file locally and apply the diff against that json file, then read the entire json file, empty the DB and repopulate the DB from the patched index file. Did I get this right? If so, certainly an interesting approach that avoids lots of issues we've been having. How's the performance with that? When downloading a full index, getting it into the DB always takes quite some time with official F-Droid.

**Comment by machiav3lli** [comment]
> Are you verifying the signature of entry.jar and the hashes of the files in there?

Not currently, as I still have to be convinced by doing this for a non-jar. Do you have an opinion on this?

> Then it looks like you keep a full index JSON file locally and apply the diff against that json file, then read the entire json file, empty the DB and repopulate the DB from the patched index file. Did I get this right? If so, certainly an interesting approach that avoids lots of issues we've been having. How's the performance with that? When downloading a full index, getting it into the DB always takes quite some time with official F-Droid.

The full index should be downloaded once, except when the client goes weeks/months long without syncing (or having a ROM/app that aggressively cleans cache of all apps). Otherwise, the only con are the couple megabytes that are kept in cache (which I would argue for devices of the last 10 years are no big issue) and maybe the full parsing (which I see acceptable considering that Sync is the heaviest processing load which a client would do). I've been testing this for couple of weeks now (the community too) and had no issues with the IoD dailies or other nightly repos (so over 14 syncs on the same repo), so I think it's stable enough to bring to the public as default.

**Comment by grote** [comment]
> >  Are you verifying the signature of entry.jar and the hashes of the files in there?
> 
> Not currently, as I still have to be convinced by doing this for a non-jar. Do you have an opinion on this?

Yes, in my book it is absolutely critical for security that you verify the hashes linked from `entry.jar`. The latter is the root of trust signed by the private repo key. So you need to verify that signature before touching the entry/index. In `entry.jar` each file is listed with a sha256 hash (so covered by the repo key signature). If you don't verify that hash, you are only relying on the security of the TLS HTTP connection. If this is broken, e.g. by a malicious CA, an attacker can substitute the full index and the diffs with arbitrary content since knowledge of the repo key isn't needed in NeoStore. For example, they can replace a popular app in a popular repo with a maliciously modified app and you wouldn't notice. Then same goes for APK downloads which also have a hash to verify. Though there the Android security model at least protects updates, but not initial installs.
